index.js

```diff
@@ -13,47 +14,25 @@ function fastifyEarlyHints (fastify, opts, next) {
     warn: opts.warn
   }
 
-  function earlyHints (reply) {
-    const promiseBuffer = []
-    const serialize = function (c) {
-      let message = `HTTP/1.1 103 Early Hints${CRLF}`
-      for (let i = 0; i < c.length; i++) {
-        message += `${formatEntry(c[i], formatEntryOpts)}${CRLF}`
-      }
-      return `${message}${CRLF}`
-    }
-
-    return {
-      close: async function () {
-        if (promiseBuffer.length) {
-          await Promise.all(promiseBuffer)
-        }
-      },
-      add: function (content) {
-        const p = new Promise(resolve => {
-          if (reply.raw.socket) {
-            reply.raw.socket.write(serialize(content), 'utf-8', resolve)
-          } else {
-            reply.raw.write(serialize(content), 'utf-8', resolve)
-          }
-        })
-        promiseBuffer.push(p)
-        return p
-      }
+  const serialize = function (c) {
+    let message = ''
+    for (let i = 0; i < c.length; i++) {
+      message += `${formatEntry(c[i], formatEntryOpts)}${CRLF}`
     }
+    return `${earlyHintsHeader}${message}${CRLF}`
   }
 
-  function onRequest (request, reply, done) {
-    reply.eh = earlyHints(reply)
-    done()
-  }
-
-  async function onSend (request, reply, payload) {
-    await reply.eh.close()
-  }
-
-  fastify.addHook('onRequest', onRequest)
-  fastify.addHook('onSend', onSend)
+  fastify.decorateReply('writeEarlyHints', function (earlyHints) {
+    const raw = this.raw
+    const serialized = serialize(earlyHints)
+    return new Promise(resolve => {
+      if (raw.socket) {
+        raw.socket.write(serialized, 'utf-8', resolve)
```

Comment:
note that the callback of `write()` could have an `err` as the first argument, which unfortunately means we should reject the promise.